[v1.10] Expose memory mapping; Make memfile dump optional

ValentaTomas opened the pull request:
To gain more control over memfile handling during snapshotting, we modify the snapshot logic to optionally skip dumping the memfile.

Also, we add info about the memory regions when requesting instance info, so we can copy the memory directly.

This is fully compatible with FC v1.10.

This PR is for viewing code changes only and won't be merged.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds GET /memory and /memory/mappings APIs for resident/empty page bitmaps and region mappings, and makes snapshot memory dump optional via mem_file_path.
> 
> - **API**:
>   - Add `GET /memory` and `GET /memory/mappings` endpoints, parsing, and responses returning `MemoryResponse` (resident/empty bitmaps) and `MemoryMappingsResponse`.
>   - Extend `InstanceInfo` with optional `memory_regions`; wire responses in `ParsedRequest::convert_to_response`.
>   - Update Swagger (`firecracker.yaml`) with new paths and definitions.
> - **VMM/VM**:
>   - Introduce `GuestMemoryRegionMapping` and methods to gather mappings and memory info using `mincore` and zero-page detection; validate uniform page size.
>   - Make snapshot memfile optional: `CreateSnapshotParams.mem_file_path: Option<PathBuf>` and dump memory only when provided.
> - **Security/Seccomp**:
>   - Allow `mincore` syscall in seccomp profiles (x86_64, aarch64).
> - **Tests**:
>   - Add unit/integration tests for new endpoints (pre-boot unsupported, post-boot success) and snapshot changes; include a non-CI memory benchmark.
> - **Tooling**:
>   - Add build/upload scripts, Makefile targets, `.tool-versions`; update `.gitignore` to include `.env`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6c05dd52688398847ba8def797275f6fd9211db4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

ValentaTomas committed:
Expose memory mapping; Add optional memfile dump

ValentaTomas committed:
Merge

ValentaTomas committed:
Fix compile errors

ValentaTomas commented:
We might also make the snapfile optional/return it as content instead.

ValentaTomas committed:
Remove required field from spec

ValentaTomas committed:
Fix parameter

ValentaTomas committed:
Fix type

ValentaTomas committed:
Add upload script

ValentaTomas committed:
Parse without python

ValentaTomas committed:
Split endpoints

ValentaTomas committed:
Fix build

ValentaTomas committed:
Rollback deps

ValentaTomas committed:
Fix build script

ValentaTomas committed:
Switch to 7 hash chars everywhere

ValentaTomas committed:
Remove g

jakubno commented:
bugbot run

jakubno commented on scripts/build.sh:
```diff
@@ -0,0 +1,17 @@
+
+set -euo pipefail
+
+# The format will be: v<major>.<minor>.<patch>_g<commit_hash> — e.g. v1.7.2_g8bb88311
```
should g be here?
```suggestion
# The format will be: v<major>.<minor>.<patch>_<commit_hash> — e.g. v1.7.2_g8bb88311
```

jakubno commented on Makefile:
```diff
@@ -0,0 +1,12 @@
+	./scripts/upload.sh $(GCP_PROJECT_ID)
+
+.PHONY: build-and-upload
+make build-and-upload: build upload
```
```suggestion
build-and-upload: build upload
```

jakubno commented on src/firecracker/src/api_server/request/memory.rs:
```diff
@@ -0,0 +1,39 @@
+
+pub(crate) fn parse_get_memory() -> Result<ParsedRequest, RequestError> {
+    METRICS.get_api_requests.instance_info_count.inc();
+    Ok(ParsedRequest::new_sync(VmmAction::GetMemory))
```
this is probably okay for now

ValentaTomas committed:
Cleanup

bchalios committed:
ci: remove dependency changes test

This is an optional test on the Firecracker side and most of the times
it's ignored (when valid dependency changes happen). Having this fail
blocks our fc-versions releases.

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
api: implement API for dirty memory

Implement API /memory/dirty which returns a bitmap tracking dirty guest
memory. The bitmap is structured as a vector of u64, so its length is:
total_number_of_pages.div_ceil(64).

Pages are ordered in the order of pages as reported by /memory/mappings.

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
feat: enable write-protection on guest memory

UFFD provides an API to enable write-protection for memory ranges
tracked by a userfault file descriptor. Detailed information can be
found here: https://docs.kernel.org/admin-guide/mm/userfaultfd.html.

To use the feature, users need to register the memory region with
UFFDIO_REGISTER_MODE_WP. Then, users need to enable explicitly
write-protection for sub-ranges of the registered region.

Writes in pages within write-protected memory ranges can be handled in
one of two ways. In synchronous mode, writes in a protected page will
cause kernel to send a write protection event over the userfaultfd.
In asynchronous mode, the kernel will automatically handle writes to
protected pages by clearing the write-protection bit. Userspace can
later observe the write protection bit by looking into the corresponding
entry of /proc/<pid>/pagemap.

This commit, uncoditionally, enables write protection for guest memory
using the asynchronous mode.

!NOTE!: asynchronous write protection requires (host) kernel version 6.7
or later).

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>
